zkstack_cli/crates/zkstack/src/commands/dev/commands/v28_precompiles.rs

```diff
@@ -0,0 +1,293 @@
+
+    let protocol_version = block_details.protocol_version.with_context(|| {
+        format!(
+            "Protocol version not present for block {}",
+            next_l2_block.as_u64()
+        )
+    })?;
+    anyhow::ensure!(
+        protocol_version >= ProtocolVersionId::Version28,
+        "THe block does not yet contain the v28 (Precompiles) upgrade"
+    );
+
+    Ok(())
+}
+
+pub async fn check_chain_readiness(
+    l1_rpc_url: String,
+    l2_rpc_url: String,
+    gw_rpc_url: String,
+    l2_chain_id: u64,
+    gw_chain_id: u64,
+) -> anyhow::Result<()> {
+    let l1_provider = get_ethers_provider(&l1_rpc_url)?;
+
+    let l2_client = get_zk_client(&l2_rpc_url, l2_chain_id)?;
+
+    let gw_client = get_ethers_provider(&gw_rpc_url)?;
+    let gw_client_l2 = get_zk_client(&gw_rpc_url, gw_chain_id)?;
+
+    // L1
+    let diamond_proxy_addr = l2_client.get_main_l1_contract().await?;
+
+    let zkchain = ZkChainAbi::new(diamond_proxy_addr, gw_client.clone());
+    let batches_committed = zkchain.get_total_batches_committed().await?.as_u32();
+    let batches_verified = zkchain.get_total_batches_verified().await?.as_u32();
+
+    verify_next_batch_new_version(batches_committed, &l2_client).await?;
+    verify_next_batch_new_version(batches_verified, &l2_client).await?;
+
+    // GW
+    let diamond_proxy_addr = gw_client_l2.get_main_l1_contract().await?;
+
+    let gw_zkchain = ZkChainAbi::new(diamond_proxy_addr, l1_provider.clone());
+    let gw_batches_committed = gw_zkchain.get_total_batches_committed().await?.as_u32();
+    let gw_batches_verified = gw_zkchain.get_total_batches_verified().await?.as_u32();
+
+    verify_next_batch_new_version(gw_batches_committed, &gw_client_l2).await?;
+    verify_next_batch_new_version(gw_batches_verified, &gw_client_l2).await?;
+
+    Ok(())
+}
+
+pub async fn fetch_chain_info(
+    upgrade_info: &V28UpgradeInfo,
+    args: &V28PrecompilesUpgradeArgsInner,
+) -> anyhow::Result<FetchedChainInfo> {
+    // Connect to the L1 Ethereum network
+    let l1_provider = get_ethers_provider(&args.l1_rpc_url)?;
+    let chain_id = U256::from(args.chain_id);
+
+    let bridgehub = BridgehubAbi::new(upgrade_info.bridgehub_addr, l1_provider.clone());
+    let zkchain_addr = bridgehub.get_zk_chain(chain_id).await?;
+    if zkchain_addr == Address::zero() {
+        anyhow::bail!("Chain not present in bridgehub");
+    }
+
+    let zkchain = ZkChainAbi::new(zkchain_addr, l1_provider.clone());
+
+    let chain_admin_addr = zkchain.get_admin().await?;
+    let l1_asset_router_proxy = bridgehub.asset_router().await?;
+
+    // Repeat for GW
+
+    let gw_client = get_ethers_provider(&args.gw_rpc_url)?;
+
+    let gw_bridgehub = BridgehubAbi::new(upgrade_info.gw_bridgehub_addr, gw_client.clone());
+    let gw_zkchain_addr = gw_bridgehub.get_zk_chain(chain_id).await?;
+    if gw_zkchain_addr == Address::zero() {
+        anyhow::bail!("Chain not present in gateway bridgehub");
```

Comment:
you dont need to bail here, a chain can be present only on L1 and never migrate to GW